packages/deployments/contracts/contracts/messaging/connectors/wormhole/BaseWormhole.sol

```diff
@@ -0,0 +1,116 @@
+
+    // Check that the VAA hasn't already been processed (replay protection)
+    require(!processedWhMessages[deliveryHash], "already processed");
+
+    // Add the VAA to processed messages so it can't be replayed
+    // you can alternatively rely on the replay protection
+    // of something like transferWithPayload from the Token Bridge module
+    processedWhMessages[deliveryHash] = true;
+
+    _processMessageFrom(fromWormholeFormat(sourceAddress), payload);
+  }
+
+  /**
+   * @dev calculcate gas to call `receiveWormholeMessages` on target chain
+   * https://github.com/wormhole-foundation/wormhole/blob/main/ethereum/contracts/relayer/deliveryProvider/DeliveryProvider.sol
+   */
+  function quoteEVMDeliveryPrice(uint256 gasLimit) public view returns (uint256 cost) {
+    (cost, ) = IWormholeRelayer(relayerAddress).quoteEVMDeliveryPrice(MIRROR_CHAIN_ID, 0, gasLimit);
+  }
+
+  // ============ Private fns ============
+  /**
+   * @dev Asserts the sender of a cross domain message
+   */
+  function _verifySender(address _mirrorConnector, address _expected) internal pure returns (bool) {
+    return _mirrorConnector == _expected;
+  }
+
+  // DO NOT override _processMessage, should revert from `Connector` class. All messages must use the _processMessageFrom function
+  /**
+   * @notice This function is called to handle incoming messages. Should store the latest
+   * root generated on the l2 domain.
+   */
+  function _processMessageFrom(address sender, bytes memory _data) internal virtual;
+
+  /**
+   * @dev send message via wormhole.
+   * https://book.wormhole.com/technical/evm/relayer.html#sending-messages
+   */
+  function _sendMessage(
+    address _mirrorConnector,
+    address _refund,
+    bytes memory _data,
+    bytes memory _encodedData
+  ) internal {
+    // Should always be sending a merkle root
+    require(_data.length == 32, "!data length");
+
+    //calculate cost to deliver message
+    uint256 gasLimit = _getGasFromEncoded(_encodedData);
+    uint256 deliveryCost = quoteEVMDeliveryPrice(gasLimit);
+    require(deliveryCost == msg.value, "!msg.value");
+
+    // publish delivery request
+    IWormholeRelayer(relayerAddress).sendPayloadToEvm{value: deliveryCost}(
+      MIRROR_CHAIN_ID,
+      _mirrorConnector,
+      _data,
+      0,
+      gasLimit,
+      MIRROR_CHAIN_ID, // refundChain
+      _refund // refundAddress
+    );
+  }
+
+  function fromWormholeFormat(bytes32 whFormatAddress) internal pure returns (address) {
+    require(uint256(whFormatAddress) >> 160 == 0, "!evm address");
+    return address(uint160(uint256(whFormatAddress)));
+  }
+
+  /**
+   * @notice Using Wormhole relayer (AMB), the gas is provided to `sendMessage` as an encoded uint
+   */
+  function _getGasFromEncoded(bytes memory _encodedData) internal view returns (uint256 _gas) {
+    // Should include gssas info in specialized calldata
+    require(_encodedData.length == 32, "!encoded data length");
```

Comment:
in other connectors, this input validation is done at the top level within `_sendMessage` I believe. we should make sure this is consistent with existing code

Reply:
ahh it's the same as in other contracts already, resolving